File: compiler/feedback/rank.go
package feedback

import (
	"slices"
	"wipple/database"
)

type Rank int

const (
	RankSyntax Rank = iota
	RankNames
	RankCustom
	RankConflicts
	RankBounds
	RankUnknown
	RankPlaceholders
)

const MaxRank Rank = RankPlaceholders

func sort(items []FeedbackItem) []FeedbackItem {
	// Reduce noise by preferring lower-rank items
	minRank := MaxRank
	for _, item := range items {
		if item.Rank < minRank {
			minRank = item.Rank
		}
	}
	items = slices.DeleteFunc(items, func(item FeedbackItem) bool {
		return item.Rank > minRank
	})

	slices.SortStableFunc(items, func(left FeedbackItem, right FeedbackItem) int {
		leftSpan := database.GetSpanFact(left.On[0])
		rightSpan := database.GetSpanFact(right.On[0])

		return leftSpan.Start.Line - rightSpan.Start.Line
	})

	return items
}

File: compiler/src/feedback/rank.rs
use crate::{database::Db, feedback::FeedbackItem};

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum FeedbackRank {
    Syntax,
    Names,
    Custom,
    Conflicts,
    Bounds,
    Unknown,
    Placeholders,
}

pub fn sort_by_rank(db: &Db, items: &mut Vec<FeedbackItem>) {
    // Reduce noise by preferring lower-rank items
    let min_rank = items.iter().map(|item| item.rank).min();

    if let Some(min_rank) = min_rank {
        items.retain(|item| item.rank == min_rank);
    }

    items.sort_by_key(|item| {
        let (node, _) = &item.location;
        db.span(node)
    });
}
